Renaming all properties, structs and methods to match the API naming

remi-stripe opened the pull request:
This PR ensures that all properties in the library matches the official API. Names are not shortened anymore making it easier to associated a given parameter to its API counterpart.

We can't list all the changes one by one as those are extensive but here are the ones likely to impact most integrations:
* All properties with the word url in it now have the same case such as `BusinessURL`.
* All properties with the word four in it now use a digit such as `DynamicLast4` or `SSNLast4`.
* Renamed confusing properties such as `Account` on `BankAccount` becoming `AccountNumber` to avoid confusing with its Connect counterpart which is now named `Account` instead of `AccountID`.
* Properties such as `MonthAnchor` or constants such as `Month` are now renamed to match the corresponding parameter/value in the API `MonthlyAnchor` and `Monthly`.
* `Meta` and `Live` have been renamed to `Metadata` and `Livemode` on all classes
* `Values` on List objects has been renamed to `Data`
* Some classes have been renamed. 
  * `Owner` is now `AdditionalOwner`,
  * `Transaction` is now `BalanceTransaction` with all constants and corresponding classes renamed to match this
  * references to `Tx` were also replaced by `BalanceTransaction` in classes, properties and constants
  * `Fee` and `FeeRefund` have been renamed to `ApplicationFee` and `ApplicationFeeRefund` along with all related classes, properties and constants.
  * `Sub` has been renamed to `Subscription` along with all related classes, properties and constants.
* Card parameters have been renamed such as `City` becoming `AddressCity` and `Month` becoming `ExpMonth.
* Errors codes have been prefixed with `ErrorCode` such as `ErrorCodeCardDeclined` or `ErrorCodeIncorrectZip`.
* Event properties are now named `Object` and `PreviousAttributes` and methods `GetObjectValue()` and `GetPreviousValue()`.
* Pagination properties have been renamed to `StartingAfter`, `EndingBefore` and `HasMore`.
* `AddMeta()` has been renamed to `AddMetadata()`, `Expand()` to `AddExpand()`

ob-stripe commented:
> We can see it touches a lot of fields and makes the PR hard to review because of identation.

Protip: you can add `?w=1` at the end of the URL to ignore whitespace changes.

https://github.com/stripe/stripe-go/pull/459/files?w=1

brandur-stripe commented:
Just to double-check: this one is all set for review right Remi?

remi-stripe commented:
@brandur-stripe Not quite yet. You wanted to fix the `bool` representation in the library first. There are also a few pending question I've shared with you internally.

remi-stripe commented:
@brandur-stripe I think this is ready to review. I will update the description to list all changes that we had in that library.

As discussed internally, I will separately open a PR that lets us move all bool in params structs to pointers.

remi-stripe commented:
Okay rebased on the latest master as a few things had changed.

brandur-stripe commented on dispute.go:
```diff
@@ -67,78 +67,78 @@ type DisputeListParams struct {
-	URL     string `json:"url"`
+	Created  int64  `json:"created"`
+	ID       string `json:"id"`
+	MimeType string `json:"mime_type"`
```
Should we do `MIMEType` (**M**ultipurpose **I**nternet **M**ail **E**xtensions).

remi-stripe replied:
fixed!

brandur-stripe commented on invoiceitem.go:
```diff
@@ -33,23 +33,23 @@ type InvoiceItem struct {
 	Proration    bool              `json:"proration"`
 	Quantity     int64             `json:"quantity"`
-	Sub          string            `json:"subscription"`
+	Subscription string            `json:"subscription"`
```
Since we're about to break it anyway, should we make `Subscription` expandable here?

remi-stripe replied:
fixed!

brandur-stripe commented on order.go:
```diff
@@ -118,10 +118,10 @@ type OrderListParams struct {
-	Paid      int64 `json:"paid"`
-	Returned  int64 `json:"returned"`
+	Canceled int64 `json:"canceled"`
+	Fulfiled int64 `json:"fulfiled"`
```
I think we should leave this as `Fulfilled` and fix it server side at some point instead.

No point in cargo culting too hard.

brandur-stripe commented on payout.go:
```diff
@@ -81,31 +81,31 @@ type PayoutListParams struct {
+	Amount                    int64               `json:"amount"`
+	ArrivalDate               int64               `json:"arrival_date"`
+	BalanceTransaction        *BalanceTransaction `json:"balance_transaction"`
+	BankAccoun                *BankAccount        `json:"bank_account"`
```
Typo here. Missing a "t" on the end.

brandur-stripe commented on subscription.go:
```diff
@@ -0,0 +1,133 @@
+package stripe
```
I see the rational here, but I wonder if we should leave this as `sub.go` so that you can always match the name of a Go file with the name of the corresponding package? (`sub/`)

remi-stripe replied:
Agreed, I just renamed those back.

brandur-stripe commented:
Great work Remi. Can't wait for this to come in!

This is definitely going to be a little painful, but even so, probably time to rip the bandaid off. I left a few comments above, but otherwise LGTM.

remi-stripe commented:
@brandur-stripe Okay, I addressed all remaining comments I think. Can you let me know if you have any other feedback?

This one will have to be merged after we merge https://github.com/stripe/stripe-go/pull/507 into it but I think we can get this done this week!

brandur-stripe commented on payout.go:
```diff
@@ -81,32 +81,32 @@ type PayoutListParams struct {
+	ArrivalDate               int64               `json:"arrival_date"`
+	Automatic                 bool                `json:"automatic"`
+	BalanceTransaction        *BalanceTransaction `json:"balance_transaction"`
+	BankAccoun                *BankAccount        `json:"bank_account"`
```
Misspelling: `BankAccount`.

remi-stripe replied:
ugh, nice catch, fixed.

brandur-stripe commented on recipienttransfer.go:
```diff
@@ -57,30 +57,30 @@ type RecipientTransferDestination struct {
+	Recipient           *Recipient                   `json:"recipient"`
+	Reversals           *ReversalList                `json:"reversals"`
+	Reversed            bool                         `json:"reversed"`
+	SourceTransation    *BalanceTransactionSource    `json:"source_transaction"`
```
Misspelling: `SourceTransaction`.

remi-stripe replied:
ugh, nice catch, fixed.

brandur-stripe commented:
@remi-stripe Nice! I took one more pass; I can't guarantee I got everything, but only found some minor stuff. Otherwise, this LGTM.

remi-stripe commented:
Rebased on the latest master. I had to squash all commits into one readable long one because there were too many merge conflicts on small ones and it was easy to miss something and introduce a regression

brandur-stripe commented:
> Rebased on the latest master. I had to squash all commits into one readable long one because there were too many merge conflicts on small ones and it was easy to miss something and introduce a regression

Nice Remi! What do you say that we try to do the work to get this merged ASAP early next week ... before you get too tired to continue this project ;)

I'm not sure if you had a particular timeline in mind, but it seems like we're going to have to pull the trigger on this at some point. We're also sitting at v29.* right now, and v30.* would be a nice round number of a big breaking change like this one :)

remi-stripe commented:
@brandur-stripe I'd love to do that. This can't be merged as is, there's the other PR waiting where I'm blocked for now though. It's unlikely to be cleaned up by early next week though and then I'm OOO for a bit.

brandur-stripe commented:
> This can't be merged as is, there's the other PR waiting where I'm blocked for now though. It's unlikely to be cleaned up by early next week though and then I'm OOO for a bit.

Ah k :/ Well in that case let's just try to get it on the board as soon as we can.

remi-stripe commented:
I've had my eyes on version 30.0.0 for a while though so I'll try to get this done!

kerak19 commented:
Hello. Any updates on this PR? Is there any approximate date when it will be merged?

brandur-stripe commented:
@kerak19 Sorry about the delayed response here. We're not exactly sure when this will come in, but Remi and I are meeting a little later today, and I'm hopeful that we can get this in within the next couple weeks.

remi-stripe commented:
Moved to https://github.com/stripe/stripe-go/pull/544 as the bigger PR
